Update expo docs (#1038)
clerk/clerk-docs: docs/references/expo/custom-auth-flows/expo-impersonation.mdx

## Before you start

1. Follow [Clerk's Expo quickstart](/docs/quickstarts/expo) to ensure your project is configured properly.
1. Because this example uses [`expo-router`](https://docs.expo.dev/router/installation/)'s experimental API routes, you must follow the [Expo docs](https://docs.expo.dev/router/reference/api-routes/#deployment) to deploy your routes to a domain. Doing so allows Clerk to access to your `CLERK_SECRET_KEY`, which you must pass to your impersonation endpoint. 
1. You must configure your Clerk instance to enable **Email and Password** authentication:
    1. Navigate to the Clerk Dashboard in the navigation sidebar, go to **User & Authentication > [Email, Phone, and Username](https://dashboard.clerk.com/last-active?path=user-authentication/email-phone-username)**.
    1. Ensure that **Email address** is required, and **Username** is not required.
    1. Ensure that **Password** is required.
1. Install the following dependencies:

>     ```bash {{ filename: "terminal" }}
>     npm i expo-linking && npm i -D @clerk/types
>     ```

Author: ````suggestion
<CodeBlockTabs options={["npm", "yarn", "pnpm"]}>
```bash {{ filename: "terminal" }}
npm i expo-linking && npm i @clerk/types
```

```bash {{ filename: "terminal" }}
yarn add expo-linking && yarn add @clerk/types
```

```bash {{ filename: "terminal" }}
pnpm add expo-linking && pnpm add @clerk/types
```
</CodeBlockTabs>
````

---
description: Learn how to build custom Impersonation into your Expo application
---

# Impersonation in Expo

<Callout type="warning">
The following flows are built assuming your Clerk app is using the default instance settings. If you have changed your instance settings, you may need to make changes to the examples.
</Callout>

User Impersonation is a powerful feature to support your customers with. Currently, from the Clerk Dashboard, [you can impersonate your users](https://clerk.com/docs/custom-flows/user-impersonation#impersonate-a-user-from-the-clerk-dashboard) and see through their eyes inside of your web applications. In this example, we’ll learn how to build impersonation into your mobile application through [Clerk’s custom flows](https://clerk.com/docs/custom-flows/overview). 

> The example we’ll build will still need some work before it goes into your app. For example, we will build impersonation directly into the dashboard, whereas you should build your impersonation into an admin only dashboard, where all of your users are listed & can be interacted with. You will want to restrict this so **only employees of your company or those who should have access to other user's information** can use this implementation.

Author: ```suggestion
> [!WARNING]
> This is a basic example that foregoes best practices for the sake of brevity and shouldn't be copied one-for-one into a production Expo application. For example, you should build impersonation into an admin-only dashboard that **only authorized users** can access.
```

I'm trimming this down and making it a warning note

## Before you start

> Before continuing, please go through the [Expo Quickstart](https://clerk.com/docs/quickstarts/expo) and utilize the repo as a starting point for this example.
> 
> ```bash
> git clone https://github.com/clerk/clerk-expo-quickstart
> ```

Author: ````suggestion
1. Follow [Clerk's Expo quickstart](/docs/quickstarts/expo) to ensure your project is configured properly.
````

I'm glad you added this point about them needing to follow the quickstart. I'll add it to the other expo custom flows in this PR 👍.

I cut the line about cloning the repo because we don't tend to point people to example repos outside of the quickstarts (as far as I can tell). I also personally think if cloning the repo isn't _required_, best to cut it so the guide only has everything they absolutely need

> 
> Since we will be using `expo-router`'s experimental API routes, you will need to follow the [Expo docs](https://docs.expo.dev/router/reference/api-routes/#deployment) to deploy your routes to a domain. This is so we have access to our `CLERK_SECRET_KEY` that we need to pass to our impersonation endpoint. 

Author: ```suggestion
1. Because this example uses [`expo-router`](https://docs.expo.dev/router/installation/)'s experimental API routes, you must follow the [Expo docs](https://docs.expo.dev/router/reference/api-routes/#deployment) to deploy your routes to a domain. Doing so allows Clerk to access to your `CLERK_SECRET_KEY`, which you must pass to your impersonation endpoint. 
```

- Markdown auto-numbers lists if every number is the same, so it's easiest to just make each number 1. That way if you delete an item or swap two items' places you don't have to renumber the list manually
- Removed "we" and "our"
- Added link to expo-router docs

Why does deploying your routes to a domain allow Clerk to access the secret key? Might be worth explaining here. What do you think?

> 
> For this project, you need to configure a few settings from the Clerk dashboard. 
> 
> 1. Configure your Clerk instance to enable Email and Password 
>     1. Navigate to the Clerk Dashboard and go to **User & Authentication > [Email, Phone, and Username](https://dashboard.clerk.com/last-active?path=user-authentication/email-phone-username)** in the sidebar menu.
>     2. Ensure that **Email address** is required, and **Username** is not required
>     3. Ensure that **Password** is required

Author: ```suggestion
1. You must configure your Clerk instance to enable **Email and Password** authentication:
    1. Navigate to the Clerk Dashboard and go to **User & Authentication > [Email, Phone, and Username](https://dashboard.clerk.com/last-active?path=user-authentication/email-phone-username)** in the sidebar menu.
    1. Ensure that **Email address** is required, and **Username** is not required
    1. Ensure that **Password** is required
1. Install the following dependencies:
    ```bash
    npm i expo-linking && npm i -D @clerk/types
    ```
```
- Adding this item to the numbered list defined with the edits above, and changing the numbers to `1`
- Moving the dependency installation instructions into the "before you start" section

## Setup and protect your Dashboard route

> To start things off, let’s install some dependencies we’ll need:
> 
> ```bash
> npm i expo-linking && npm i -D @clerk/types
> ```
> 

Author: ```suggestion
```
Moved this section up to "before you start"

            user_id: userId, // this is the user ID of the use you're going to impersonate,
            actor: {
              sub: actorId, // this is the ID of the impersonator,
            },
          }),
        });

        const data = await res.json();

        setActor(data);
      }
    }

    generateAndSetToken();
  }, []);

  return actor;
}
```

Now our endpoint will call Clerk’s [create actor token](https://clerk.com/docs/reference/backend-api/tag/Actor-Tokens#operation/CreateActorToken) endpoint at `/actor_tokens` and pass in the Clerk secret key for authorization.

```tsx
// generateActorToken+api.tsx

export async function POST(request: Request) {
  const body: { user_id: string; actor: { sub: string } } =
    await request.json();

  const { user_id, actor } = body;

  const res = await fetch("https://api.clerk.com/v1/actor_tokens", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${process.env.CLERK_SECRET_KEY}`,
      "Content-type": "application/json",
    },
    body: JSON.stringify({
      user_id,
      actor: {
        sub: actor.sub,
      },
    }),
  });

  const data = await res.json();

  return Response.json(data);
}
```

> ## Get the Impersonated user’s data
> 
> The second hook will get back user data on the impersonator. You can use this to display UI only the impersonator will see for specific actions, like showing who is currently impersonating and matching the session to the right user for logout. 
> 
> ```tsx
> // useImpersonatedUser

Author: ````suggestion
### Create a hook to get impersonated user data

Next, create a second hook that will fetch data about the impersonator. You can use this to display UI only the impersonator will see for specific actions, like showing who is currently impersonating and matching the session to the right user for logout. 

```tsx "app/(dashboard)/index.tsx"
````

I changed the title a little to drop the word "the" so it's not as long in the ToC but if that title is inaccurate we can keep it as "Create a hook to get the impersonated user's data"

function useImpersonatedUser(
  actorSub: string,
  setImpersonator: React.Dispatch<React.SetStateAction<string | UserDataJSON>>
) {
  React.useEffect(() => {
    const getImpersonatedUser = async () => {
      const res = await fetch(`/getImpersonatedUser`, {
        method: "POST",
        body: JSON.stringify({
          impersonator_id: actorSub,
        }),
      });

      const data = await res.json();

      setImpersonator(data);

      getImpersonatedUser();
    };
  }, [actorSub]);
}
```

> With our hook setup we now can create our API route that will call Clerk’s retrieve user endpoint, `/users`, and get back the Impersonated user’s full `User` object.
> 
> ```tsx
> // getImpersonatedUser+api.tsx

Author: ````suggestion
### Create an API route to retrieve user data

With your hook setup you can now create an API route that will call [Clerk’s retrieve user endpoint](https://clerk.com/docs/reference/backend-api/tag/Users#operation/GetUserList), `/users`, and get back the Impersonated user’s full [`User` object](/docs/references/javascript/user/user#user-object).

```tsx filename="app/getImpersonatedUser+api.tsx"
````

export async function POST(request: Request) {
  const body: { impersonator_id: string } = await request.json();

  const { impersonator_id } = body;
  const res = await fetch(`https://api.clerk.com/v1/users/${impersonator_id}`, {
    headers: {
      Authorization: `Bearer ${process.env.CLERK_SECRET_KEY}`,
    },
  });
  const data = await res.json();

  return Response.json(data);
}
```

> Now that we have the functionality for generating an impersonation session & getting their user data, we have a few more functions to complete that will get the `ticket` from the URL generated by our `useImpersonation` hook, pass that to our `signIn` function to create a new impersonated session, and allow for logout of the impersonated or the impersonator. 
> 
> ## Generate an impersonation session
> 
> Let’s create our functions that work together to generate a new impersonate session, `extractTicket` and `impersonateUser`
> 
> ```tsx
>   // use this function to get the ticket ID from the response of our generateActorToken API

Author: ````suggestion

### Create functions to generate an impersonated session

Now that you have the functionality for generating an impersonated session & getting the impersonated user's data, there are a few more functions to complete that will:

1. Get the `ticket` from the URL generated by your `useImpersonation` hook
1. Pass the `ticket` to our `signIn` function to create a new impersonated session, allowing you to sign out the impersonated user or impersonator.

The following code should be placed in the `Page` component of your `app/(dashboard)/index.tsx` file:

```tsx filename="app/(dashboard)/index.tsx"
  // use this function to get the ticket ID from the response of our generateActorToken API
````

- Cleaned things up a bit
- Broke a big sentence into a list so it's easier to read

  function extractTicketValue(input: string): string | undefined {
    const index = input.indexOf("ticket=");
    if (index !== -1) {
      return input.slice(index + 7);
    }
    return undefined;
  }
	
	// passing the ticket ID to our `signIn.create` function will create a new impersonated session that we can set to the active session
  async function impersonateUser() {
    if (!isLoaded) return;

    if (typeof actorRes?.url === "string") {
      const ticket = extractTicketValue(actorRes.url);

      if (ticket) {
        try {
          const { createdSessionId } = await signIn.create({
            strategy: "ticket",
            ticket,
          });

          await setActive({ session: createdSessionId });
          await user?.reload();

          router.replace("/dashboard");
        } catch (err) {
          console.log(err);
        }
      }
    }
  }
```

> Finally, we have our functionality used to signout. Usually this can be as simple as calling Clerk’s `signOut` function, but since we’re handling multiple sessions we’ll need to add some checks to make sure we’re signing out of the right session. 
> 
> ```tsx

Author: ````suggestion

### Create a hook for signing out

Finally, create a hook for signing out. Usually this can be as simple as calling Clerk’s [`signOut`](https://clerk.com/docs/pr/1038/references/javascript/clerk/clerk#sign-out) function, but since to handling multiple sessions you must add some checks to ensure you're signing out of the right session. 

```tsx filename="app/(dashboard)/index.tsx"
````

  const onSignOutPress = async (sessionId: string) => {
    try {
      if (isLoaded && sessions && sessions?.length > 0) {
        const noActiveSessions = sessions.filter(
          (session) => session.user?.id !== user?.id
        );
        await setActive({ session: noActiveSessions[0].id });
      }
      const redirectUrl = Linking.createURL("/dashboard", { scheme: "myapp" });
      await signOut({
        sessionId,
      });
      router.replace(redirectUrl);
    } catch (err: any) {}
  };
```

> ## Bringing it all together
> 
> The Dashboard page will now look like this, excluding the API routes;

Author: ```suggestion
### Full example

The full code for the dashboard page will resemble the following, excluding the API routes:
```
